ethcore/src/miner/stratum.rs

```diff
@@ -0,0 +1,193 @@
+			Err(e) => {
+				warn!(target: "stratum", "submit_work: invalid nonce ({:?})", e);
+				return;
+			}
+		};
+
+
+		let pow_hash = match H256::from_str(&payload[1]) {
+			Ok(pow_hash) => pow_hash,
+			Err(e) => {
+				warn!(target: "stratum", "submit_work: invalid hash ({:?})", e);
+				return;
+			}
+		};
+
+		let mix_hash = match H256::from_str(&payload[2]) {
+			Ok(mix_hash) => mix_hash,
+			Err(e) => {
+				warn!(target: "stratum", "submit_work: invalid mix-hash ({:?})", e);
+				return;
+			}
+		};
+
+		trace!(target: "stratum", "submit_work: Decoded: nonce={}, pow_hash={}, mix_hash={}", nonce, pow_hash, mix_hash);
+		let client = self.client.upgrade().unwrap();
+		let miner = self.miner.upgrade().unwrap();
+
+		let seal = vec![encode(&mix_hash).to_vec(), encode(&nonce).to_vec()];
+		if let Err(e) = miner.submit_seal(&*client, pow_hash, seal) {
+			warn!(target: "stratum", "submit_work error: {:?}", e);
+		};
+	}
+}
+
+impl StratumJobDispatcher {
+	/// New stratum job dispatcher given the miner and client
+	fn new(miner: &Arc<Miner>, client: &Arc<Client>) -> StratumJobDispatcher {
+		StratumJobDispatcher {
+			seed_compute: Mutex::new(SeedHashCompute::new()),
+			last_work: RwLock::new(None),
+			client: Arc::downgrade(client),
+			miner: Arc::downgrade(miner),
+		}
+	}
+
+	/// Serializes payload for stratum service
+	fn payload(&self, pow_hash: H256, difficulty: U256, number: u64) -> String {
+		// TODO: move this to engine
+		let target = Ethash::difficulty_to_boundary(&difficulty);
+		let seed_hash = &self.seed_compute.lock().get_seedhash(number);
+		let seed_hash = H256::from_slice(&seed_hash[..]);
+		format!(
+			r#"["0x{}","0x{}","0x{}","0x{:x}"]"#,
+			pow_hash.hex(), seed_hash.hex(), target.hex(), number
+		)
+	}
+
+	fn with_core<F, R>(&self, f: F) -> Option<R> where F: Fn(Arc<Client>, Arc<Miner>) -> Option<R> {
+		self.client.upgrade().and_then(|client| self.miner.upgrade().and_then(|miner| (f)(client, miner)))
+	}
+}
+
+/// Wrapper for dedicated stratum service
+pub struct Stratum {
+	dispatcher: Arc<StratumJobDispatcher>,
+	base_dir: String,
+	stop: ::devtools::StopGuard,
+}
+
+pub enum Error {
+	Nano(nanoipc::SocketError),
+}
+
+impl From<nanoipc::SocketError> for Error {
+	fn from(socket_err: nanoipc::SocketError) -> Error { Error::Nano(socket_err) }
+}
+
+impl super::work_notify::NotifyWork for Stratum {
+	#[allow(unused_must_use)]
+	fn notify(&self, pow_hash: H256, difficulty: U256, number: u64) {
+		nanoipc::generic_client::<RemoteWorkHandler<_>>(&format!("ipc://{}/ipc/parity-stratum.ipc", self.base_dir))
+			.and_then(|client| {
+				client.push_work_all(
+					self.dispatcher.payload(pow_hash, difficulty, number)
+				).unwrap_or_else(
+					|e| warn!(target: "stratum", "Error while pushing work: {:?}", e)
+				);
+				*self.dispatcher.last_work.write() = Some((pow_hash, difficulty, number));
+				Ok(client)
+			})
+			.map_err(|e| warn!(target: "stratum", "Can't connect to stratum service: {:?}", e));
```

Comment:
Yes, but if it's consistently down it would be much better to know about it upstream - that warning could be turned into an error and `notify` could return `Result<(), Error>` which could be then handled properly (even if it's only the warning)